### ISSUE TITLE
Bliss-ful events?

### ISSUE THREAD

**Issue by jonathantneal** [LeaVerou/bliss#90]
I’d love to see simple, chain-able events. This is a proposal I’ve been kicking around. Do you think Bliss would be a good home for it?

**Chaining events**:
```js
$element.on('keypress').then(function (event) {
	// log the key code each time a key is pressed
	console.log(event.keyCode);
}).then(function (event) {
	// then log whether the shift key is pressed
	console.log(event.shiftKey);
});
```
---
**Filtering events**:
```js
$element.on('keypress').filter(function (event) {
	// pass the keypress event forward if the shift key is active
	return event.shiftKey;
}).then(function (event) {
	// then log the key code (if the shift key is active)
	console.log(event.keyCode);
});
```
---
**Mapping events**:
```js
$element.on('keypress').map(function (event) {
	// pass the key code forward
	return event.keyCode;
}).then(function (keyCode) {
	// then log the key code
	console.log(keyCode);
});
```
---
**Removing events**:
```js
$element.on('keypress').then(function (event) {
	// log the key code
	console.log(event.keyCode);
}).until(function (event) {
	// then remove the listener if the shift key is pressed
	return event.shiftKey;
});
```

```js
// drop any keypress events
$element.off('keypress');
```
---
**Triggering events**:
```js
$element.on('custom-event').forEach(function (event) {
	console.log(event); // includes { anything: 'you want' }
});

$element.do('custom-event', {
	anything: 'you want'
});
```

**Comment by kurtextrem** [comment]
First 3 seem good to me :+1: (is `until` a good expression? Would `.finally` be better?)

I think `.do` is a bit unusual, it would be better to use `.trigger` (like jQuery)

**Comment by dperrymorrow** [comment]
```$.fire``` and ```$.events``` seems to cover the situations you are describing. 

http://blissfuljs.com/docs.html#fn-fire
http://blissfuljs.com/docs.html#fn-events

**Comment by jonathantneal** [comment]
@kurtextrem, Thanks! Yes, or even `fire` (as is brought up in the following comment). I’ll add that `until` is borrowed from RxJS. IMO, `finally` might imply that it is to be the last event to fire regardless of what it returns, whereas `until` suggests the return value might unsubscribe the chain, as in "*do this `until` this method returns truthy*”.

@dperrymorrow, `fire` certainly covers `do`, but I’m not sure how either of those methods cover any of the event chaining or transforming.

**Comment by dperrymorrow** [comment]
yes you are correct ```$.events``` does not have "chain able" events. But what i was referring to was that you could accomplish the same thing that your code demonstrates using ```$.events```.

**your example:**
```javascript
$element.on('keypress').then(function (event) {
    // log the key code each time a key is pressed
    console.log(event.keyCode);
}).then(function (event) {
    // then log whether the shift key is pressed
    console.log(event.shiftKey);
});
```

**using events:**
```javascript
$element._.events({
  keyup: function () {
    console.log(event.keyCode);
    console.log(event.shiftKey);
  }
});
```

So what i meant was you can accomplish the same thing with ```$.events``` but what you are proposing is a syntactic sugar for that functionaly IMO. Ill leave it up to @LeaVerou if that is something she sees as a good idea for Bliss.

Thanks for the input and ideas! Much appreciated.

**Comment by LeaVerou** [comment]
I used to have a method that let you bind events and returned a promise, which I removed before releasing Bliss, as I wasn't sure it was useful (and also wasn't sure what to do with multiple elements). Perhaps something like that would help here?
Otherwise, something like this would probably require a lot of code, for functionality that I'm not sure is as commonly needed. If we can come up with little code to address the use cases (even with a different syntax), that would be cool. I'll leave this open for a bit in case anyone has ideas.
